File: system/fake/ImageBuilder.java
package common.system.fake;

import common.io.assets.Admin.StaticPermitted;
import common.system.VImg;
import common.system.files.FileData;
import common.system.files.VFile;
import common.util.Data;

import java.io.ByteArrayInputStream;
import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.util.function.Supplier;

public abstract class ImageBuilder<T> {

	@StaticPermitted(StaticPermitted.Type.ENV)
	public static ImageBuilder<?> builder;

	public static VImg toVImg(byte[] bs) {
		return new VImg(Data.err(() -> builder.build(bs)));
	}

	public static VImg toVImg(File f) {
		return new VImg(Data.err(() -> builder.build(f)));
	}

	public final FakeImage build(byte[] bs) throws IOException {
		return build((Supplier<InputStream>) () -> new ByteArrayInputStream(bs));
	}

	public abstract FakeImage build(File f) throws IOException;

	public final FakeImage build(FileData fd) throws IOException {
		return build((Supplier<InputStream>) fd::getStream);
	}

	public abstract FakeImage build(Supplier<InputStream> sup) throws IOException;

	public abstract FakeImage build(T o);

	public final FakeImage build(VFile vf) throws IOException {
		return build(vf.getData());
	}

	public final VImg toVImg(T t) {
		return new VImg(Data.err(() -> build(t)));
	}

	public abstract boolean write(FakeImage o, String fmt, Object out) throws IOException;

}

File: system/fake/ImageBuilder.kt
package common.system.fake

import common.io.assets.Admin
import common.io.assets.Admin.StaticPermitted
import java.io.IOException

abstract class ImageBuilder {
    /**
     * List of parsed Classes (to `Supplier<InputStream>`):
     *  * `VFile`
     *  * `FileData`
     *  * `byte[]`
     * <hr></hr>
     * List of processed Classes:
     *  * `File`
     *  * `Supplier<InputStream>`
     * <hr></hr>
     * List of transfered Classes:
     *  * `FakeImage` return directly
     *  * `BufferedImage` PC only
     *  * `BitMap` android only
     *
     * @throws IOException
     */
    @Throws(IOException::class)
    abstract fun build(o: Any?): FakeImage?
    @Throws(IOException::class)
    abstract fun write(o: FakeImage, fmt: String?, out: Any?): Boolean

    companion object {
        @StaticPermitted(Admin.StaticPermitted.Type.ENV)
        var builder: ImageBuilder? = null
    }
}
